06_puppy/brvishnu92/main.go

```diff
@@ -0,0 +1,104 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+var out io.Writer = os.Stdout
+
+type Puppy struct {
+	id    uint32
+	breed string
+	color string
+	value string
+}
+
+type Storer interface {
+	CreatePuppy(*Puppy)
+	ReadPuppy(id uint32) *Puppy
+	UpdatePuppy(puppy *Puppy)
+	DeletePuppy(id uint32) bool
+	HasPuppy(id uint32) bool
+}
+
+type MapStore map[uint32]Puppy
+
+type SynStore struct {
+	store *sync.Map
+}
+
+func (ms MapStore) CreatePuppy(pup *Puppy) {
+	if !ms.HasPuppy(pup.id) {
+		ms[pup.id] = *pup
+	}
+}
+func (ms MapStore) ReadPuppy(id uint32) *Puppy {
+	if puppy, has := ms[id]; has {
+		return &puppy
+	}
+	return nil
+}
+
+func (ms MapStore) UpdatePuppy(puppy *Puppy) {
+	if ms.HasPuppy(puppy.id) {
+		ms[puppy.id] = *puppy
+	}
+}
+
+func (ms MapStore) DeletePuppy(id uint32) bool {
+	if ms.HasPuppy(id) {
+		delete(ms, id)
+		return true
+	}
+	return false
+}
+func (ms MapStore) HasPuppy(id uint32) bool {
+	_, has := ms[id]
+	return has
+}
+
+func (ss SynStore) CreatePuppy(pup *Puppy) {
+	if !ss.HasPuppy(pup.id) {
```

Comment:
The point of using sync.Map is to protect your code from race conditions (make it "thread/goroutine safe").
In your case HasPuppy reads before Store updates. Another go routine could in the meantime alter the result of HasPuppy.
To mitigate this you can use [Mutex](https://golang.org/pkg/sync/#Mutex) or [RWMutex](https://golang.org/pkg/sync/#RWMutex) - use lock at beginning of method and unlock at end where instructions need to be executed together.
Same for all other methods that use two or more access to the underlying syncmap